src/libstd/sys/unix/ext/net.rs

```diff
@@ -330,124 +329,42 @@ impl fmt::Debug for SocketAddr {
 /// ```
 #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
 #[derive(Debug)]
-pub struct SocketAncillary {
-    buffer: Vec<u8>,
+pub struct SocketAncillary<'a> {
+    buffer: &'a mut [u8],
+    length: usize,
     truncated: bool,
 }
 
-impl SocketAncillary {
-    /// Create a new empty ancillary data with a capacity of zero.
+impl<'a> SocketAncillary<'a> {
+    /// Create an ancillary data with the given buffer.
     ///
     /// # Example
     ///
     /// ```no_run
     /// # #![allow(unused_mut)]
     /// #![feature(unix_socket_ancillary_data)]
     /// use std::os::unix::net::SocketAncillary;
-    /// let mut ancillary = SocketAncillary::new();
+    /// let mut ancillary_buffer = [0; 128];
+    /// let mut ancillary = SocketAncillary::new(&mut ancillary_buffer[..]);
     /// ```
     #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
-    pub fn new() -> Self {
-        SocketAncillary::with_capacity(0)
+    pub fn new(buffer: &'a mut [u8]) -> Self {
+        unsafe {
+            libc::memset(buffer.as_mut_ptr().cast(), 0, buffer.len());
+        }
+        SocketAncillary { buffer, length: 0, truncated: false }
     }
 
-    /// Increase the capacity for `n` file descriptors.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the new capacity overflows `usize`.
-    ///
-    /// # Example
-    ///
-    /// ```no_run
-    /// #![feature(unix_socket_ancillary_data)]
-    /// use std::os::unix::net::{UnixStream, SocketAncillary};
-    /// use std::io::IoSliceMut;
-    ///
-    /// fn main() -> std::io::Result<()> {
-    ///     let sock = UnixStream::connect("/tmp/sock")?;
-    ///
-    ///     let mut fds = [0; 8];
-    ///     let mut ancillary = SocketAncillary::new();
-    ///     ancillary.add_capacity_for_fds(8);
-    ///
-    ///     let mut buf = [1; 8];
-    ///     let mut bufs = &mut [IoSliceMut::new(&mut buf[..])][..];
-    ///     sock.recv_vectored_with_ancillary(bufs, &mut ancillary)?;
-    ///
-    ///     ancillary.get_fds(&mut fds[..]);
-    ///     Ok(())
-    /// }
-    /// ```
+    /// Returns the capacity of the buffer.
     #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
-    pub fn add_capacity_for_fds(&mut self, n: usize) -> &mut Self {
-        add_capacity_to_ancillary_data::<RawFd>(&mut self.buffer, n);
-        self
```

Comment:
completely getting rid of these methods doesn't seem great either since it pushes all the size calculation onto the user. At least some helper methods or constants to do the most common calculations would be good.

Reply:
I tried to figured out what would be the best replacement for these functions. `const` functions are not stable yet in Rust and constants are not portable. At least the macro `CMSG_SPACE` should be used to calculate the size. 

So a helper function should be used, which call `CMSG_SPACE`. @the8472, what's your opinion?

Reply:
I think this is another indicator that it's not a good idea to put the buffer management on the API user. We should continue the discussion about this as noted in https://github.com/rust-lang/rust/pull/69864#pullrequestreview-406168075

But sure, if it can't be const and there's agreement that we keep the `&mut [u8]` passing then a normal helper function would be needed.

Reply:
> const functions are not stable yet in Rust

Huh? `const fn` has been stable for some time. There are limitations to what they can do, but they're still plenty useful.

Reply:
Thanks for the info, I did not know that. Is it possible to call an unsafe function, which is not `const`, like [CMSG_SPACE](https://docs.rs/libc/0.2.69/libc/fn.CMSG_SPACE.html)?

Reply:
No, there is no way to call a non-const function from a constant. There is no way to evaluate it at compile time. For example calling the native `CMSG_SPACE` would be wrong when cross-compiling.